MuonAnalysis/MuonAssociators/plugins/MuonFSRProducer.cc

```diff
@@ -0,0 +1,229 @@
+
+    }
+   static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("packedPFCandidates",edm::InputTag("packedPFCandidates"))->setComment("packed pf candidates where to look for photons");
+    desc.add<edm::InputTag>("slimmedElectrons",edm::InputTag("slimmedElectrons"))->setComment("electrons to check for footprint");
+    desc.add<edm::InputTag>("muons",edm::InputTag("slimmedMuons"))->setComment("collection of muons to correct for FSR ");
+    desc.add<double>("muonPtMin",20.)->setComment("minimum pt of the muon to look for a near photon");
+    desc.add<double>("muonEtaMax",2.4)->setComment("max eta of the muon to look for a near photon");
+    desc.add<double>("photonPtMin",2.0)->setComment("minimum photon Pt");
+    desc.add<double>("deltaROverEt2Max",0.05)->setComment("max ratio of deltsR(mu,photon) over et2 of the photon");
+    desc.add<double>("isolation",2.0)->setComment("relative isolation cut");
+
+    descriptions.add("MuonFSRProducer", desc);
+  } 
+  ~MuonFSRProducer() override {}
+  
+  //static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+private:
+  
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  double computeRelativeIsolation(const pat::PackedCandidate & photon,
+				  const pat::PackedCandidateCollection& pfcands,
+				  const double & isoConeMax,
+				  const double & isoConeMin) const;
+
+// ----------member data ---------------------------
+  const edm::EDGetTokenT<pat::PackedCandidateCollection> pfcands_;
+  const edm::EDGetTokenT<pat::ElectronCollection> electrons_;
+  const edm::EDGetTokenT<edm::View<reco::Muon>> muons_;
+  float ptCut;
+  float etaCut;
+  float photonPtCut;
+  float drEtCut;
+  float isoCut;
+
+
+};
+
+void MuonFSRProducer::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+
+  using namespace std;
+
+  edm::Handle<pat::PackedCandidateCollection> pfcands;
+  iEvent.getByToken(pfcands_, pfcands);
+  edm::Handle<edm::View<reco::Muon>> muons;
+  iEvent.getByToken(muons_, muons);
+  edm::Handle<pat::ElectronCollection> electrons;
+  iEvent.getByToken(electrons_, electrons);
+
+  std::vector<int> muonMapping(muons->size(),-1);
+  auto fsrPhotons = std::make_unique<std::vector<pat::GenericParticle>>();
+  // loop over all muons
+  for (auto muon = muons->begin(); muon != muons->end(); ++muon){
+
+    
+    int photonPosition = -1;
+    double distance_metric_min = -1;
+    // minimum muon pT
+    if (muon->pt() < ptCut) continue;
+    // maximum muon eta
+    if (fabs(muon->eta() > etaCut)) continue;
+
+    // for each muon, loop over all pf cadidates
+    for(auto iter_pf = pfcands->begin(); iter_pf != pfcands->end(); iter_pf++){
+      auto const & pc = *iter_pf;
+
+      
+      // consider only photons
+      if (abs(pc.pdgId()) != 22) continue;
+      //cout<<"ID pass!"<<endl;
```

Comment:
I guess these comments should be removed

Reply:
done